registry/storage/driver/s3-aws/s3.go

```diff
@@ -1164,28 +1116,44 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
 			})
 		}
 
-		sort.SliceStable(walkInfos, func(i, j int) bool { return walkInfos[i].FileInfoFields.Path < walkInfos[j].FileInfoFields.Path })
-
 		for _, walkInfo := range walkInfos {
-			err := f(walkInfo)
+			// skip any results under the last skip directory
+			if prevSkipDir != "" && strings.HasPrefix(walkInfo.Path(), prevSkipDir) {
+				continue
+			}
 
-			if err == storagedriver.ErrSkipDir {
-				if walkInfo.IsDir() {
-					continue
-				} else {
-					break
+			dir := filepath.Dir(walkInfo.Path())
+			if dir != prevDir {
+				prevDir = dir
+				walkDirInfo := storagedriver.FileInfoInternal{
+					FileInfoFields: storagedriver.FileInfoFields{
+						IsDir: true,
+						Path:  dir,
+					},
 				}
-			} else if err != nil {
-				retError = err
-				return false
-			}
+				err := f(walkDirInfo)
+				*objectCount++
 
-			if walkInfo.IsDir() {
-				if err := d.doWalk(ctx, objectCount, *walkInfo.prefix, prefix, f); err != nil {
+				if err != nil {
+					if err == storagedriver.ErrSkipDir {
+						prevSkipDir = dir
+						continue
+					}
 					retError = err
 					return false
 				}
 			}
+
+			err := f(walkInfo)
+			*objectCount++
+
+			if err != nil {
+				if err == storagedriver.ErrSkipDir {
```

Comment:
Is it possible that `ErrSkipDir` is returned for an actual object? In that case, what was the previous behavior? Seems to me like we would want to do the same thing as when this err is returned for a "directory".

Reply:
Following what is documented, 
```
	// Walk traverses a filesystem defined within driver, starting
	// from the given path, calling f on each file.
	// If the returned error from the WalkFn is ErrSkipDir and fileInfo refers
	// to a directory, the directory will not be entered and Walk
	// will continue the traversal.  If fileInfo refers to a normal file, processing stops
	Walk(ctx context.Context, path string, f WalkFn) error
```
and based on the impl of `WalkFallback` I'm interpreting `ErrSkipDir` on a file to mean "break" aka graceful stop. So actually, should be returning `false` to stop the S3 list.